pkg/controller/clusterdeployment/clusterdeployment_controller.go

```diff
@@ -0,0 +1,720 @@
+	if err != nil && !errors.IsNotFound(err) {
+		return fmt.Errorf("error retrieving master machineset %s/%s: %v", clusterDeployment.Namespace, machineSetName, err)
+	}
+
+	if machineSet == nil {
+		// machine set does not exist, it needs to be created
+		machineSet, err = buildMasterMachineSet(clusterDeployment, clusterVersion)
+		c.logger.Debugf("Just built machineset: %#v", machineSet)
+		if err != nil {
+			return fmt.Errorf("error building machineSet from clusterDeployment %s/%s: %v", clusterDeployment.Namespace, clusterDeployment.Name, err)
+		}
+		_, err := c.capiClient.ClusterV1alpha1().MachineSets(clusterDeployment.Namespace).Create(machineSet)
+		if errors.IsConflict(err) {
+			// machine set was created elsewhere, no need to retry
+			clustoplog.WithClusterDeployment(c.logger, clusterDeployment).Debugf("conflict creating master machineset, skipping")
+			return nil
+		}
+		if err != nil {
+			return fmt.Errorf("error creating machineSet %s/%s: %v", machineSet.Namespace, machineSet.Name, err)
+		}
+		return nil
+	}
+
+	// master machine set exists, make sure it reflects the current cluster deployment master machine set spec
+	machineSetConfig := masterMachineSetConfig(clusterDeployment)
+	if machineSetConfig == nil {
+		return fmt.Errorf("cluster deployment %s/%s does not have a master machineSetConfig", clusterDeployment.Namespace, clusterDeployment.Name)
+	}
+	specProviderConfig, err := controller.MachineProviderConfigFromMachineSetConfig(machineSetConfig, &clusterDeployment.Spec, clusterVersion)
+	if err != nil {
+		return fmt.Errorf("cannot create a machine providerconfig from machineset config for cluster deployment %s/%s: %v", clusterDeployment.Namespace, clusterDeployment.Name, err)
+	}
+	if string(machineSet.Spec.Template.Spec.ProviderConfig.Value.Raw) != string(specProviderConfig.Raw) {
+		updatedMachineSet := machineSet.DeepCopy()
+		updatedMachineSet.Spec.Template.Spec.ProviderConfig.Value = specProviderConfig
+		_, err := c.capiClient.ClusterV1alpha1().MachineSets(updatedMachineSet.Namespace).Update(updatedMachineSet)
+		if err != nil {
+			return fmt.Errorf("error updating machineset %s/%s: %v", machineSet.Namespace, machineSet.Name, err)
+		}
+	}
+	return nil
+}
+
+func buildMasterMachineSet(clusterDeployment *clustop.ClusterDeployment, clusterVersion *clustop.ClusterVersion) (*capi.MachineSet, error) {
+	machineSetConfig := masterMachineSetConfig(clusterDeployment)
+	if machineSetConfig == nil {
+		return nil, fmt.Errorf("cluster deployment %s/%s does not have a master machine set config", clusterDeployment.Namespace, clusterDeployment.Name)
+	}
+
+	machineSet := &capi.MachineSet{}
+	machineSet.Name = masterMachineSetName(clusterDeployment.Name)
+	machineSet.Namespace = clusterDeployment.Namespace
+	machineSet.Labels = clusterDeployment.Labels
+	if machineSet.Labels == nil {
+		machineSet.Labels = make(map[string]string)
+	}
+	machineSet.Labels[clusterDeploymentLabel] = clusterDeployment.Name
```

Comment:
Should we also add a cluster label until we get something into upstream? My new AWS PR is relying on the cluster name label during delete. Perhaps I should be changing that to cluster deployment once this merges though.